File: C++/Day10.cpp
#include "AH.h"

namespace Day10
{

	int score1(const char c)
	{
		switch(c) {
			case ')':
				return 3;
			case ']':
				return 57;
			case '}':
				return 1197;
			case '>':
				return 25137;
			default:
				return 0;
		} 
	}

	int64_t score2(const std::string s)
	{
		int64_t score = 0;
		for (int i = s.length() - 1; i >= 0; --i)
		{
			score *= 5;
			switch(s[i]) {
				case '(':
					score += 1; break;
				case '[':
					score += 2; break;
				case '{':
					score += 3; break;
				case '<':
					score += 4; break;
				default:
					return -1;
			} 			
		}
		return score;
	}

	bool closing(const char l, const char r)
	{
		return ((l == '(' && r == ')') || (l == '{' && r == '}') ||
		        (l == '[' && r == ']') || (l == '<' && r == '>'));
	}

	std::pair<char, std::string> parseString(const std::string& s)
	{
		std::string stck = "";
		const std::string bras = "({[<";
		for (auto c : s)
		{
			if (bras.find(c) != std::string::npos)
				stck += c;
			else if (closing(stck.back(), c))
				stck.pop_back(); 
			else
				return std::make_pair(c, "");
		}

		return std::make_pair('X', stck);
	}

	int Run(const std::string& filename)
	{
		auto inputLines = AH::ReadTextFile(filename);

		int part1 = 0;
		std::vector<int64_t> part2Arr;
		int goodCount = 0;
		for (auto s : inputLines)
		{
			auto [a, b] = parseString(s);
			if (a == 'X') {
				++goodCount;
				part2Arr.push_back(score2(b));
			}

			part1 += score1(a);
		}
		std::sort(part2Arr.begin(), part2Arr.end());

		AH::PrintSoln(10, part1, part2Arr[goodCount / 2]);

		return 0;
	}

}

File: Go/Day10.go
package main

import AH "./adventhelper"

import (
	"sort"
)

func score1(r rune) int {
	if r == ')' {
		return 3
	}
	if r == ']' {
		return 57
	}
	if r == '}' {
		return 1197
	}
	if r == '>' {
		return 25137
	}

	return 0
}

func score2(s string) (score int) {
	score = 0
	for _, v := range s {
		score *= 5
		if v == '(' {
			score += 1
		}
		if v == '[' {
			score += 2
		}
		if v == '{' {
			score += 3
		}
		if v == '<' {
			score += 4
		}
	}

	return
}

func closing(l rune, r rune) bool {
	return (l == '(' && r == ')') || (l == '{' && r == '}') ||
	       (l == '[' && r == ']') || (l == '<' && r == '>')
}

func illegal(s string) (rune, string) {
	stack := ""
	for _, r := range s {
		if AH.ContainsChar("({<[", r) {// bra
			stack = string(r) + stack
		} else if (closing(AH.FirstRune(stack), r)) {
			stack = AH.TrimFirstRune(stack)
		} else {
			return r, "" // found bad character
		}
	}
	// reached the end of the string
	return 'X', stack
}

func main() {
	inputLines, _ := AH.ReadStrFile("../input/input10.txt")

	part1 := 0
	part2arr := []int{}
	goodCount := 0

	for _, s := range inputLines {
		a,b := illegal(s)

		part1 += score1(a)
		if a == 'X' {
			goodCount += 1
			part2arr = append(part2arr, score2(b))
		}
	}
	sort.Sort(sort.IntSlice(part2arr))

	AH.PrintSoln(10, part1, part2arr[goodCount / 2])

	return
}
